File: lib/teto_bot/interactions/leaderboard.ex
defmodule TetoBot.Interactions.Leaderboard do
  @moduledoc """
  Handles leaderboard-related functionality for Discord interactions.
  """

  require Logger

  alias Nostrum.Api
  alias TetoBot.Accounts
  alias TetoBot.Interactions.Responses

  @api_timeout 2000
  @task_timeout 1500

  @spec handle_leaderboard(Nostrum.Struct.Interaction.t(), integer()) :: :ok | Api.error()
  @doc """
  Handles the leaderboard command interaction.
  """
  def handle_leaderboard(interaction, guild_id) do
    case Accounts.get_leaderboard(guild_id) do
      {:ok, []} ->
        Responses.success(interaction, "No one has earned intimacy with Teto in this guild yet!")

      {:ok, entries} ->
        build_and_send_leaderboard(interaction, guild_id, entries)

      {:error, reason} ->
        Logger.error("Failed to fetch leaderboard for guild #{guild_id}: #{inspect(reason)}")

        Responses.error(
          interaction,
          "Failed to retrieve the leaderboard. Please try again later."
        )
    end
  end

  @spec build_and_send_leaderboard(Nostrum.Struct.Interaction.t(), integer(), list()) ::
          :ok | Api.error()
  defp build_and_send_leaderboard(interaction, guild_id, entries) do
    user_ids = Enum.map(entries, & &1.user_id)

    case fetch_user_data_concurrently(guild_id, user_ids) do
      {:ok, {members_map, users_map}} ->
        leaderboard_text = build_leaderboard_text(entries, members_map, users_map)
        Responses.success(interaction, leaderboard_text)

      {:error, :timeout} ->
        Responses.error(
          interaction,
          "The leaderboard is temporarily unavailable due to slow Discord API response. Please try again later."
        )

      {:error, reason} ->
        Logger.error("Failed to fetch user data for leaderboard: #{inspect(reason)}")

        Responses.error(
          interaction,
          "Failed to retrieve the leaderboard. Please try again later."
        )
    end
  end

  defp fetch_user_data_concurrently(guild_id, user_ids) do
    parent = self()

    task =
      Task.async(fn ->
        # Fetch guild members and users concurrently
        member_task = Task.async(fn -> Api.Guild.members(guild_id, limit: 1000) end)
        user_tasks = Enum.map(user_ids, &Task.async(fn -> Api.User.get(&1) end))

        # Await results
        {:ok, members} = Task.await(member_task, @api_timeout)
        users = Task.await_many(user_tasks, @api_timeout)

        # Process results
        members_map =
          members
          |> Enum.map(fn m -> {m.user_id, m} end)
          |> Map.new()

        users_map =
          users
          |> Enum.map(fn {:ok, u} -> {u.id, u} end)
          |> Map.new()

        send(parent, {:user_data_fetched, {:ok, {members_map, users_map}}})
      end)

    receive do
      {:user_data_fetched, result} ->
        Task.shutdown(task)
        result
    after
      @task_timeout ->
        Task.shutdown(task, :brutal_kill)
        {:error, :timeout}
    end
  end

  defp build_leaderboard_text(entries, members_map, users_map) do
    leaderboard_entries =
      entries
      |> Enum.with_index(1)
      |> Enum.map(fn {%{user_id: user_id, intimacy: intimacy}, rank} ->
        display_name = get_display_name(user_id, members_map, users_map)
        "#{rank}. #{display_name} - #{intimacy}"
      end)

    leaderboard = Enum.join(leaderboard_entries, "\n")

    """
    **Teto's Intimacy Leaderboard (Top 10)**

    #{leaderboard}
    """
  end

  defp get_display_name(user_id, members_map, users_map) do
    # Try nickname from guild member first
    case Map.get(members_map, user_id) do
      %{nick: nick} when not is_nil(nick) ->
        nick

      _ ->
        # Fall back to user global name or username
        case Map.get(users_map, user_id) do
          %{global_name: global_name} when not is_nil(global_name) -> global_name
          %{username: username} when not is_nil(username) -> username
          _ -> "User##{user_id}"
        end
    end
  end
end

File: src/commands/leaderboard.ts
import {
  type ChatInputCommandInteraction,
  MessageFlags,
  SlashCommandBuilder,
  type User,
} from 'discord.js'
import { Effect, Either, Runtime } from 'effect'
import { buildLeaderboardEmbed } from '../embeds/leaderboard'
import { ApiService, ChannelService, type MainLive } from '../services'
import type { ApiError, LeaderboardEntry } from '../services/api/client'

const API_TIMEOUT = 2000

export const data = new SlashCommandBuilder()
  .setName('leaderboard')
  .setDescription("View Teto's intimacy leaderboard for this server")

/**
 * Create a promise that rejects after a timeout
 */
function createTimeoutPromise<T>(timeoutMs: number): Promise<T> {
  return new Promise((_, reject) => {
    setTimeout(() => reject(new Error('Timeout')), timeoutMs)
  })
}

/**
 * Fetch user data from Discord API with timeout handling
 * Refactored to avoid guild_members intent by only fetching user data
 */
async function fetchUserDataConcurrently(
  interaction: ChatInputCommandInteraction,
  userIds: string[]
): Promise<Map<string, User>> {
  try {
    // Only fetch user data, no guild members
    const userPromises = userIds.map((userId) =>
      interaction.client.users.fetch(userId).catch((error) => {
        console.warn(`Failed to fetch user ${userId}:`, error)
        return null
      })
    )

    // Race against timeout
    const dataPromise = Promise.all(userPromises)
    const timeoutPromise = createTimeoutPromise<(User | null)[]>(API_TIMEOUT)

    const users = await Promise.race([dataPromise, timeoutPromise])

    // Build map from the results
    const usersMap = new Map<string, User>()
    for (const user of users) {
      if (user) {
        usersMap.set(user.id, user)
      }
    }

    return usersMap
  } catch (error) {
    if (error instanceof Error && error.message === 'Timeout') {
      throw new Error('TIMEOUT')
    }
    throw error
  }
}

/**
 * Build and send leaderboard with concurrent user data fetching
 */
async function buildAndSendLeaderboard(
  interaction: ChatInputCommandInteraction,
  entries: LeaderboardEntry[]
): Promise<void> {
  const userIds = entries.map((entry) => entry.userId)

  try {
    const usersMap = await fetchUserDataConcurrently(interaction, userIds)

    const leaderboardEmbed = buildLeaderboardEmbed(
      entries,
      usersMap,
      interaction.guild?.name
    )

    await interaction.reply({
      embeds: [leaderboardEmbed],
    })
  } catch (error) {
    console.error('Failed to fetch user data for leaderboard:', error)

    if (error instanceof Error && error.message === 'TIMEOUT') {
      await interaction.reply({
        content:
          'The leaderboard is temporarily unavailable due to slow Discord API response. Please try again later.',
        flags: MessageFlags.Ephemeral,
      })
    } else {
      await interaction.reply({
        content: 'Failed to retrieve the leaderboard. Please try again later.',
        flags: MessageFlags.Ephemeral,
      })
    }
  }
}

/**
 * Effect-based leaderboard operation
 */
const fetchLeaderboardEffect = (
  guildId: string,
  userId: string
): Effect.Effect<LeaderboardEntry[], ApiError, ApiService> =>
  Effect.gen(function* () {
    const apiService = yield* ApiService
    const effectApi = apiService.effectApi

    // Ensure the guild and user exist in the database first
    // This handles cases where the guild creation failed or the user is new
    yield* effectApi.discord.ensureUserGuildExists({
      userId,
      guildId,
    })

    const result = yield* effectApi.leaderboard.getIntimacyLeaderboard({
      guildId,
      limit: 10,
    })

    return result.data.leaderboard
  }).pipe(
    Effect.tapError((error) =>
      Effect.logError(
        `Failed to fetch leaderboard for guild ${guildId}: ${error.message}`
      )
    )
  )

export async function execute(
  runtime: Runtime.Runtime<never>,
  live: typeof MainLive,
  interaction: ChatInputCommandInteraction
) {
  const guildId = interaction.guildId
  const channelId = interaction.channelId
  const userId = interaction.user.id

  if (!guildId) {
    await interaction.reply({
      content: 'This command can only be used in a server.',
      flags: MessageFlags.Ephemeral,
    })
    return
  }

  const isChannelWhitelisted = await ChannelService.pipe(
    Effect.flatMap(({ isChannelWhitelisted }) =>
      isChannelWhitelisted(channelId)
    )
  ).pipe(Effect.provide(live), Runtime.runPromise(runtime))

  if (!isChannelWhitelisted) {
    await interaction.reply({
      content: 'This command can only be used in whitelisted channels.',
      flags: MessageFlags.Ephemeral,
    })
    return
  }

  // Convert Effect to Either and run it
  const program = fetchLeaderboardEffect(guildId, userId).pipe(
    Effect.either,
    Effect.provide(live)
  )
  const result = await Runtime.runPromise(runtime)(program)

  // Handle Either result
  if (Either.isLeft(result)) {
    // Error case
    await interaction.reply({
      content: 'Failed to retrieve the leaderboard. Please try again later.',
      flags: MessageFlags.Ephemeral,
    })
    return
  }

  const entries = result.right

  if (entries.length === 0) {
    await interaction.reply({
      content: 'No one has earned intimacy with Teto in this guild yet!',
    })
    return
  }

  await buildAndSendLeaderboard(interaction, entries)
}
